core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -1318,4 +1417,137 @@ private static class Frame {
+    if (sqlNode instanceof SqlSelect) {
+      fetchSqlWithItemNodes(sqlNode, sqlNodes);
+    } else if (sqlNode instanceof SqlBasicCall) {
+      fetchfromSqlBasicCall(sqlNode, sqlNodes);
+    } else if (sqlNode instanceof SqlJoin) {
+      SqlNode leftNode = ((SqlJoin) sqlNode).getLeft();
+      SqlNode rightNode = ((SqlJoin) sqlNode).getRight();
+
+      fetchSqlWithItems(leftNode, sqlNodes);
+      fetchSqlWithItems(rightNode, sqlNodes);
+    } else if (sqlNode instanceof SqlWithItem) {
+      fetchFromSqlWithItemNode(sqlNode, sqlNodes);
+    }
+  }
+
+  private void fetchfromSqlBasicCall(SqlNode sqlNode, List<SqlNode> sqlNodes) {
+    SqlNode sqlNodeOPerand = ((SqlBasicCall) sqlNode).operand(0);
+    if (sqlNodeOPerand instanceof SqlSelect) {
+      fetchSqlWithItemNodes(sqlNodeOPerand, sqlNodes);
+    } else if (sqlNodeOPerand instanceof SqlWithItem) {
+      fetchFromSqlWithItemNode(sqlNodeOPerand, sqlNodes);
+    }
+  }
+
+  private void fetchFromSqlWithItemNode(SqlNode sqlWithItem, List<SqlNode> sqlNodes) {
+    fetchSqlWithItems(((SqlWithItem) sqlWithItem).query, sqlNodes);
+    updateSqlNode(((SqlWithItem) sqlWithItem).query);
+    addSqlWithItemNode((SqlWithItem) sqlWithItem, sqlNodes);
+
+  }
+
+  private void addSqlWithItemNode(SqlWithItem sqlWithItem, List<SqlNode> sqlNodes) {
+    if (sqlNodes.isEmpty()) {
+      sqlNodes.add(sqlWithItem);
+    } else {
+      boolean status = false;
+      for (SqlNode sqlWithItemNode : sqlNodes) {
+        if (((SqlWithItem) sqlWithItemNode).name.toString()
+             .equalsIgnoreCase(sqlWithItem.name.toString())) {
+          status = true;
+          break;
+        }
+      }
+      if (!status) {
+        sqlNodes.add(sqlWithItem);
+      }
+
+    }
+  }
+
+  private SqlNode updateSqlWithNode(Result result) {
+    SqlSelect sqlSelect = null;
+    if (result.node instanceof SqlSelect) {
+      sqlSelect = (SqlSelect) result.node;
+    } else {
+      sqlSelect = wrapSelect(result.node);
+    }
+    updateSqlNode(sqlSelect);
+    return sqlSelect;
+  }
+
+  private void updateSqlNode(SqlNode sqlNode) {
+    if (sqlNode != null) {
+      SqlSelect sqlSelect = (SqlSelect) sqlNode;
+      if (sqlSelect.getFrom() instanceof SqlJoin) {
+        SqlNode leftNode = ((SqlJoin) sqlSelect.getFrom()).getLeft();
+        SqlNode rightNode = ((SqlJoin) sqlSelect.getFrom()).getRight();
+
+        //Update Left Node
+        updateNode(leftNode, sqlSelect);
+        // Update Right Node
+        updateNode(rightNode, sqlSelect);
+
+        //if left and right is join again need recursive call
+      } else if (sqlSelect.getFrom() instanceof SqlBasicCall) {
+        updateNode(sqlSelect.getFrom(), sqlSelect);
+      } else if (sqlSelect.getFrom() instanceof SqlWithItem) {
+        sqlSelect.setFrom(((SqlWithItem) sqlSelect.getFrom()).name);
+      }
+    }
+  }
+
+  private void updateNode(SqlNode sqlNode, SqlSelect selectNode) {
+    if (sqlNode instanceof SqlBasicCall) {
```

Comment:
give more relevant names and if needed a comment.

Reply:
Added Comments